### PR TITLE
Add tracking of memory allocation per operator

### PR DESCRIPTION
**Pull request by mbasmanova**
Here is a snippet of the query JSON fetched from the coordinator featuring new fields: addInput/getOutput/finishAllocation

```
  }, {
      "stageId" : 1,
      "stageExecutionId" : 0,
      "pipelineId" : 1,
      "operatorId" : 1,
      "planNodeId" : "4",
      "operatorType" : "HashAggregationOperator",
      "totalDrivers" : 408,
      "addInputCalls" : 1,
      "addInputWall" : "23.92ms",
      "addInputCpu" : "22.66ms",
      "addInputAllocation" : "3.64MB",
      "rawInputDataSize" : "0B",
      "rawInputPositions" : 0,
      "inputDataSize" : "115B",
      "inputPositions" : 1,
      "sumSquaredInputPositions" : 1.0,
      "getOutputCalls" : 820,
      "getOutputWall" : "33.97ms",
      "getOutputCpu" : "31.83ms",
      "getOutputAllocation" : "376.24kB",
      "outputDataSize" : "115B",
      "outputPositions" : 1,
      "physicalWrittenDataSize" : "0B",
      "blockedWall" : "0.00ns",
      "finishCalls" : 411,
      "finishWall" : "16.55ms",
      "finishCpu" : "15.69ms",
      "finishAllocation" : "94.47kB",
      "userMemoryReservation" : "0B",
      "revocableMemoryReservation" : "0B",
      "systemMemoryReservation" : "0B",
      "peakUserMemoryReservation" : "387.73kB",
      "peakSystemMemoryReservation" : "0B",
      "peakTotalMemoryReservation" : "387.73kB",
      "spilledDataSize" : "0B"
    }, {
```

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Comment by wenleix** [comment]
cc @viczhang861 : Can this deprecate the tagged memory pool? 😃  (discussed in https://github.com/prestodb/presto/pull/14117)

**Comment by viczhang861** [comment]
@wenleix  Briefly went through change sin this PR,  it updates allocation when an operator completes.  Tagged memory pool is updated whenever some bytes are reserved, which is more up-to-date (that is why it is expensive but useful for debugging purpose).

File: presto-main/src/main/java/com/facebook/presto/operator/OperationTimer.java [hunk 1]
```diff
@@ -53,11 +60,18 @@
         this.trackOperationCpuTime = trackOperationCpuTime;
         checkArgument(trackOverallCpuTime || !trackOperationCpuTime, "tracking operation cpu time without tracking overall cpu time is not supported");
 
+        this.trackOverallAllocation = trackOverallAllocation;
+        this.trackOperationAllocation = trackOperationAllocation;
+        checkArgument(trackOverallAllocation || !trackOperationAllocation, "tracking operation allocation without tracking overall allocation is not supported");
+        checkArgument(!trackOverallAllocation || THREAD_MX_BEAN_EX != null, "tracking allocation is not supported by this JVM");
```
Comment (viczhang861) [inline presto-main/src/main/java/com/facebook/presto/operator/OperationTimer.java]: - This is simpler to understand  :) 
     `!(trackOverallAllocation && THREAD_MX_BEAN_EX == null)`  
    `!(trackOperationAllocation && !trackOverallAllocation)` 
- Move second check up

Comment (mbasmanova) [reply presto-main/src/main/java/com/facebook/presto/operator/OperationTimer.java]: @viczhang861 Changed to

```
        this.trackOverallAllocation = trackOverallAllocation;
        this.trackOperationAllocation = trackOperationAllocation;
        if (trackOverallAllocation) {
            checkArgument(THREAD_MX_BEAN_EX != null, "tracking allocation is not supported by this JVM");
        }
        else {
            checkArgument(!trackOperationCpuTime, "tracking operation cpu time without tracking overall cpu time is not supported");
        }
```

Comment (viczhang861) [reply presto-main/src/main/java/com/facebook/presto/operator/OperationTimer.java]: Much better. I restarted a failed test.

**Comment by viczhang861** [comment]
> Everything looks good except for a nit regarding the naming "allocation timer enabled" and "per operator allocation timer enabled". The CPU version of this is a timer because it's measuring time.

I think `timer` refers to an interval, not the specific measurement type. For example, `OperationTiming` also tracks `calls`.

File: presto-main/src/main/java/com/facebook/presto/operator/OperationTimer.java [hunk 2]
```diff
@@ -53,11 +60,22 @@
         this.trackOperationCpuTime = trackOperationCpuTime;
         checkArgument(trackOverallCpuTime || !trackOperationCpuTime, "tracking operation cpu time without tracking overall cpu time is not supported");
 
+        this.trackOverallAllocation = trackOverallAllocation;
+        this.trackOperationAllocation = trackOperationAllocation;
+        if (trackOverallAllocation) {
+            checkArgument(THREAD_MX_BEAN_EX != null, "tracking allocation is not supported by this JVM");
+        }
+        else {
+            checkArgument(!trackOperationCpuTime, "tracking operation cpu time without tracking overall cpu time is not supported");
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/operator/OperationTimer.java]: `tracking operator allocations without ...`

**Comment by mbasmanova** [comment]
@aweisberg @arhimondr Ariel, Andrii, thank you for reviews. Comments addressed. CC: @rschlussel

File: presto-main/src/main/java/com/facebook/presto/operator/TaskContext.java [hunk 3]
```diff
@@ -113,6 +116,8 @@ public static TaskContext createTaskContext(
             MemoryTrackingContext taskMemoryContext,
             boolean perOperatorCpuTimerEnabled,
             boolean cpuTimerEnabled,
+            boolean perOperatorAllocationTimerEnabled,
+            boolean allocationTimerEnabled,
```
Comment (aweisberg) [inline presto-main/src/main/java/com/facebook/presto/operator/TaskContext.java]: Still says timer

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecutorFactory.java [hunk 4]
```diff
@@ -135,7 +140,9 @@ public PrestoSparkTaskExecutorFactory(
             DataSize maxSpillMemory,
             DataSize sinkMaxBufferSize,
             boolean perOperatorCpuTimerEnabled,
-            boolean cpuTimerEnabled)
+            boolean cpuTimerEnabled,
+            boolean perOperatorAllocationTimerEnabled,
+            boolean allocationTimerEnabled)
```
Comment (aweisberg) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecutorFactory.java]: Timer here as well

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecutorFactory.java [hunk 5]
```diff
@@ -149,6 +156,8 @@ public PrestoSparkTaskExecutorFactory(
         this.sinkMaxBufferSize = requireNonNull(sinkMaxBufferSize, "sinkMaxBufferSize is null");
         this.perOperatorCpuTimerEnabled = perOperatorCpuTimerEnabled;
         this.cpuTimerEnabled = cpuTimerEnabled;
+        this.perOperatorAllocationTimerEnabled = perOperatorAllocationTimerEnabled;
+        this.allocationTimerEnabled = allocationTimerEnabled;
```
Comment (aweisberg) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecutorFactory.java]: And these two

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecutorFactory.java [hunk 6]
```diff
@@ -192,6 +201,8 @@ public IPrestoSparkTaskExecutor create(
                 session,
                 perOperatorCpuTimerEnabled,
                 cpuTimerEnabled,
+                perOperatorAllocationTimerEnabled,
+                allocationTimerEnabled,
```
Comment (aweisberg) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecutorFactory.java]: Last one!

File: presto-main/src/main/java/com/facebook/presto/operator/TaskContext.java [hunk 7]
```diff
@@ -125,6 +130,8 @@ public static TaskContext createTaskContext(
                 taskMemoryContext,
                 perOperatorCpuTimerEnabled,
                 cpuTimerEnabled,
+                perOperatorAllocationTimerEnabled,
+                allocationTimerEnabled,
```
Comment (aweisberg) [inline presto-main/src/main/java/com/facebook/presto/operator/TaskContext.java]: Timer

**Comment by mbasmanova** [comment]
@aweisberg Ariel, thank you for review. I renamed the remaining timers.

**Comment by aweisberg** [comment]
LGTM! 🚢 it.

**Pull request merged**